### PR TITLE
Add state checkpointing to other training scripts

### LINKED ISSUE #666: How to resume from checkpoint.
I followed Colab for training. May I know how to resume the training.

Thanks and regards

### PR DESCRIPTION
**Pull request by pcuenca**
Bring dreambooth changes in #1668 to the rest of the training scripts.
Fixes #666.

### CODE REVIEW

File: examples/research_projects/dreambooth_inpaint/train_dreambooth_inpaint.py [hunk 1]
```diff
@@ -628,14 +647,39 @@ def collate_fn(examples):
...
     logger.info(f"  Gradient Accumulation steps = {args.gradient_accumulation_steps}")
     logger.info(f"  Total optimization steps = {args.max_train_steps}")
+    global_step = 0
+    first_epoch = 0
+
+    if args.resume_from_checkpoint:
+        if args.resume_from_checkpoint != "latest":
+            path = os.path.basename(args.resume_from_checkpoint)
+        else:
+            # Get the mos recent checkpoint
```
Comment (patrickvonplaten) [inline examples/research_projects/dreambooth_inpaint/train_dreambooth_inpaint.py:657]:
```suggestion
            # Get the most recent checkpoint
```

File: examples/research_projects/dreambooth_inpaint/train_dreambooth_inpaint.py [hunk 2]
```diff
@@ -719,6 +763,12 @@ def collate_fn(examples):
                 progress_bar.update(1)
                 global_step += 1
 
+                if global_step % args.checkpointing_steps == 0:
+                    if accelerator.is_main_process:
+                        save_path = os.path.join(args.output_dir, f"checkpoint-{global_step}")
```
Comment (patrickvonplaten) [inline examples/research_projects/dreambooth_inpaint/train_dreambooth_inpaint.py:768]: naming is good for me, since there is no need to allow the user to overwrite "checkpoint-..." with something else

File: examples/textual_inversion/textual_inversion.py [hunk 3]
```diff
@@ -543,17 +562,42 @@ def main():
...
     logger.info(f"  Gradient Accumulation steps = {args.gradient_accumulation_steps}")
     logger.info(f"  Total optimization steps = {args.max_train_steps}")
+    global_step = 0
+    first_epoch = 0
+
+    if args.resume_from_checkpoint:
+        if args.resume_from_checkpoint != "latest":
+            path = os.path.basename(args.resume_from_checkpoint)
+        else:
+            # Get the mos recent checkpoint
```
Comment (patrickvonplaten) [inline examples/textual_inversion/textual_inversion.py:572]:
```suggestion
            # Get the most recent checkpoint
```

File: examples/unconditional_image_generation/train_unconditional.py [hunk 4]
```diff
@@ -351,11 +368,36 @@ def transforms(examples):
...
 
     global_step = 0
-    for epoch in range(args.num_epochs):
+    first_epoch = 0
+
+    if args.resume_from_checkpoint:
+        if args.resume_from_checkpoint != "latest":
+            path = os.path.basename(args.resume_from_checkpoint)
+        else:
+            # Get the mos recent checkpoint
```
Comment (patrickvonplaten) [inline examples/unconditional_image_generation/train_unconditional.py:377]:
```suggestion
            # Get the most recent checkpoint
```

File: examples/research_projects/dreambooth_inpaint/train_dreambooth_inpaint.py [hunk 5]
```diff
@@ -242,6 +242,24 @@ def parse_args():
         ),
     )
     parser.add_argument("--local_rank", type=int, default=-1, help="For distributed training: local_rank")
+    parser.add_argument(
+        "--checkpointing_steps",
+        type=int,
+        default=500,
+        help=(
+            "Save a checkpoint of the training state every X updates. These checkpoints are only suitable for resuming"
```
Comment (patrickvonplaten) [inline examples/research_projects/dreambooth_inpaint/train_dreambooth_inpaint.py:250]:
```suggestion
            "Save a checkpoint of the training state every X updates. These checkpoints can be used both as final checkpoints in case they are better than the last checkpoint and are suitable for resuming"
```

The checkpoints should also be useful by themselves no? What couldn't one re-use those for inference?  

We could maybe add to the docs how to use intermediate checkpoints for inference, but it's definitely possible and should also be encouraged (see issue here: https://github.com/huggingface/diffusers/issues/732). People will just have to "build" their pipeline afterwards, but this shouldn't be too hard.

**Review by patrickvonplaten — COMMENTED** [review]
Would maybe rephrase the help message a bit and possible even add an example of how to use intermediate checkpoints for inference

**Comment by patrickvonplaten** [comment]
BTW @pcuenca , `accelerate` already merged the "only last checkpoitns" features to "main": https://github.com/huggingface/accelerate/pull/916 

Let's maybe allow people to directly use it depending on the accelerate version?

File: examples/research_projects/dreambooth_inpaint/train_dreambooth_inpaint.py (hunk 5 above)
Comment (pcuenca) [reply examples/research_projects/dreambooth_inpaint/train_dreambooth_inpaint.py:250]: Great point, I'll add it to the docs.

**Comment by pcuenca** [comment]
> BTW @pcuenca , `accelerate` already merged the "only last checkpoitns" features to "main": [huggingface/accelerate#916](https://github.com/huggingface/accelerate/pull/916)
> 
> Let's maybe allow people to directly use it depending on the accelerate version?

Yes!

**Review by patrickvonplaten — APPROVED** [review]
Perf thanks!

**Comment by sanghyun0927** [comment]
Sorry for that I am newbie in github

Just want you to know that this line may have to be changed

examples/research_projects/dreambooth_inpaint/train_dreambooth_inpaint.py

675    for epoch in range(first_epoch, args.num_epochs):
==>
675    for epoch in range(first_epoch, args.num_train_epochs):

File: examples/research_projects/dreambooth_inpaint/train_dreambooth_inpaint.py (hunk 2 above)
Comment (sanghyun0927) [reply examples/research_projects/dreambooth_inpaint/train_dreambooth_inpaint.py:769]: Sorry for that I am newbie in github

Just want you to know that this line may have to be changed

examples/research_projects/dreambooth_inpaint/train_dreambooth_inpaint.py

675 for epoch in range(first_epoch, args.num_epochs):
==>
675 for epoch in range(first_epoch, args.num_train_epochs):

**Comment by pcuenca** [comment]
I think you are right, @sanghyun0927, just opened #1885 to fix it. Thanks a lot!
